susemanager-utils/susemanager-sls/modules/roster/uyuni.py

```diff
@@ -0,0 +1,309 @@
+                PROXY_SSH_PUSH_USER,
+                "-W %s:%s" % (minion_id, ssh_push_port)
+                if not tunnel and i == len(proxies) - 1
+                else "",
+                proxy,
+            )
+        )
+        i += 1
+    if tunnel:
+        proxyCommand += (
+            "/usr/bin/ssh -i {pushKey} -o StrictHostKeyChecking=no "
+            "-o User={user} -R {pushPort}:{proxy}:{sslPort} {minion} "
+            "ssh -i {ownKey} -W {minion}:{sshPort} "
+            "-o StrictHostKeyChecking=no -o User={user} {minion}".format(
+                pushKey=PROXY_SSH_PUSH_KEY,
+                user=user,
+                pushPort=SSH_PUSH_PORT_HTTPS,
+                proxy=proxies[len(proxies) - 1],
+                sslPort=SSL_PORT,
+                minion=minion_id,
+                ownKey="{}{}".format(
+                    "/root" if user == "root" else "/home/{}".format(user),
+                    "/.ssh/mgr_own_id",
+                ),
+                sshPort=ssh_push_port,
+            )
+        )
+    proxyCommand += "'"
+
+    return [proxyCommand]
+
+
+def _getSSHMinion(
+    minion_id=None, proxies=[], tunnel=False, ssh_push_port=SSH_PUSH_PORT
+):
+    user = SSH_PUSH_SUDO_USER if SSH_PUSH_SUDO_USER else "root"
+    minion = {
+        "host": minion_id,
+        "user": user,
+        "port": ssh_push_port,
+        "timeout": SALT_SSH_CONNECT_TIMEOUT,
+    }
+    if tunnel:
+        minion.update({"minion_opts": {"master": minion_id}})
+    if proxies:
+        minion.update(
+            {
+                "ssh_options": _getSSHOptions(
+                    minion_id=minion_id,
+                    proxies=proxies,
+                    tunnel=tunnel,
+                    user=user,
+                    ssh_push_port=ssh_push_port,
+                )
+            }
+        )
+    elif tunnel:
+        minion.update(
+            {
+                "remote_port_forwards": "%d:%s:%d"
+                % (SSH_PUSH_PORT_HTTPS, COBBLER_HOST, SSL_PORT)
+            }
+        )
+
+    return minion
+
+
+def targets(tgt, tgt_type="glob", **kwargs):
+    """
+    Return the targets from the Uyuni DB
+    """
+
+    ret = {}
+
+    cache_data = cache.fetch("roster/uyuni", "minions")
+    cache_fp = cache_data.get("fp", None)
+    query = """
+        SELECT FORMAT('%s|%s|%s|%s',
+                      EXTRACT(EPOCH FROM MAX(S.modified)),
+                      COUNT(S.id),
+                      EXTRACT(EPOCH FROM MAX(SP.modified)),
+                      COUNT(SP.proxy_server_id)
+               ) AS fp
+               FROM rhnServer AS S
+               INNER JOIN suseServerContactMethod AS SSCM ON
+                     (SSCM.id=S.contact_method_id)
+               INNER JOIN suseMinionInfo AS SMI ON
+                     (SMI.server_id=S.id)
+               LEFT JOIN rhnServerPath AS SP ON
+                    (SP.server_id=S.id)
+               WHERE S.contact_method_id IN (
```

Comment:
if you are joining with the suseServerContactMethod in the from clause, you can just filter it by label, the same way it is being done in the query down in the code, with `WHERE SSCM.label IN ('ssh-push', 'ssh-push-tunnel')`

Reply:
It's done different way just because I don't need to get the actual values of `SSM.label` in this case, but only validate the cache. I can do it the same way in the query as in the next one, but according to the output of `explain ....`, the cost of the query is lower in this case, not dramatically, but it was an attempt to optimize it.

Reply:
But in this case, why are you joining in the from clause with the table `suseServerContactMethod`? I think table `suseMinionInfo` is also not used.
Do you think query can be simplefied to:
```
SELECT FORMAT('%s|%s|%s|%s',
                      EXTRACT(EPOCH FROM MAX(S.modified)),
                      COUNT(S.id),
                      EXTRACT(EPOCH FROM MAX(SP.modified)),
                      COUNT(SP.proxy_server_id)
               ) AS fp
               FROM rhnServer AS S
               LEFT JOIN rhnServerPath AS SP ON
                    (SP.server_id=S.id)
               WHERE S.contact_method_id IN (
                         SELECT SCCM.id
                         FROM suseServerContactMethod AS SCCM
                         WHERE ScCM.label IN ('ssh-push', 'ssh-push-tunnel')
                     );

```

Reply:
Then it's better to do this way:
```
SELECT FORMAT('%s|%s|%s|%s',
                      EXTRACT(EPOCH FROM MAX(S.modified)),
                      COUNT(S.id),
                      EXTRACT(EPOCH FROM MAX(SP.modified)),
                      COUNT(SP.proxy_server_id)
               ) AS fp
               FROM rhnServer AS S
               INNER JOIN suseMinionInfo AS SMI ON
                     (SMI.server_id=S.id)
               LEFT JOIN rhnServerPath AS SP ON
                    (SP.server_id=S.id)
               WHERE S.contact_method_id IN (
                         SELECT SSCM.id
                         FROM suseServerContactMethod AS SSCM
                         WHERE SSCM.label IN ('ssh-push', 'ssh-push-tunnel')
                     )
```

We don't need to drop `suseMinionInfo` from the query as it make sense only for salt systems, but on dropping this part it will invalidate the cache on changing traditional clients also.
And additionally I think it's better to extend `suseMinionInfo` by adding `created` and `modified` fields to make fingerprint sensitive for changing this table also.

Reply:
I Agree with your analysis. the proposed query looks good to me

Reply:
I just extended it slightly by appending `created` and `modified` to `suseMinionInfo`.